pkgs/development/python-modules/skein/default.nix

```diff
@@ -0,0 +1,101 @@
+{ autoPatchelfHook
+, buildPythonPackage
+, coreutils
+, fetchPypi
+, jre
+, lib
+, maven
+, protobuf
+, pythonPackages
+, stdenv
+}:
+
+buildPythonPackage rec {
+  pname = "skein";
+  version = "0.8.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0nb64p1hzshgi1kfc2jx1v9vn8b0wzs50460wfra3fsxh0ap66ab";
+  };
+
+  skeinRepo = stdenv.mkDerivation rec {
+    name = "${pname}-${version}-maven-repo";
+
+    inherit src;
+
+    nativeBuildInputs = [ maven ] ++ lib.optional stdenv.isLinux autoPatchelfHook;
+
+    buildPhase = "";
+
+    installPhase = ''
+      mkdir -p $out
+
+      archs="${
+        if stdenv.isLinux
+        then "linux-x86_32 linux-x86_64"
+        else "osx-x86_64"
+      }"
+
+      for arch in $archs
+      do
+        mvn -Dmaven.repo.local=$out dependency:get -Dartifact=com.google.protobuf:protoc:3.0.0:exe:$arch
+        mvn -Dmaven.repo.local=$out dependency:get -Dartifact=io.grpc:protoc-gen-grpc-java:1.16.0:exe:$arch
+      done
+
+      if ${ lib.boolToString stdenv.isLinux }
+      then
+        autoPatchelf $out
+      fi
+
+      # We have to use maven package here as dependency:go-offline doesn't
+      # fetch every required jar.
+      mvn -f java/pom.xml -Dmaven.repo.local=$out package
+
+      rm $(find $out -name _remote.repositories)
+      rm $(find $out -name resolver-status.properties)
+    '';
+
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+    outputHash = if stdenv.isLinux
+      then "12f0q3444qw6y4f6qsa9540a0fz4cgi844zzi8z1phqn3k4dnl6v"
+      else "0bjbwiv17cary1isxca0m2hsvgs1i5fh18z247h1hky73lnhbrz8";
+
```

Comment:
is this trying to make it a fixed output? seems odd with mkDerivation

Reply:
This is how fixed output derivations are implemented under the hood. There's no existing fetchFoo that will do this right, but it is still a normal fixed output approach.

Reply:
interesting, but I still don't see the value in defining the outputHash if nix is able to determine a derivation hash

Reply:
> I still don't see the value in defining the outputHash if nix is able to determine a derivation hash

The maven dependencies are not managed by nix. IMHO this derivation is really like a complex `fetchurl` which downloads stuff from the internet. I wouldn't want to use a fetchurl without an expected hash too. 

There are some examples in nixpkgs of the like. Most notably buildRustPackage which fetches dependencies from upstream and expects the user to provide hash to guarantee reproducibility.

Reply:
@jonringer This drv fetches from the net. It needs to be fixed output.